### PR TITLE
KAFKA-3537: Expose metrics registry

### PR DESCRIPTION
**Pull request by enothereska**

### CODE REVIEW

**Comment by enothereska** [comment]
@guozhangwang @dguy have a look when you can. Thanks.

File: streams/src/test/java/org/apache/kafka/streams/state/KeyValueStoreTestDriver.java [hunk 1]
```diff
@@ -191,6 +193,10 @@ public Sensor addLatencySensor(String scopeName, String entityName, String opera
         @Override
         public void recordLatency(Sensor sensor, long startNs, long endNs) {
         }
+        @Override
+        public Map<MetricName, ? extends Metric> metrics() {
+            return null;
```
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/state/KeyValueStoreTestDriver.java:199]: I know this is only a test class, but returning `Collections.emptyMap()` would be the nicer thing to do here

File: streams/src/test/java/org/apache/kafka/streams/state/internals/StreamThreadStateStoreProviderTest.java [hunk 2]
```diff
@@ -236,5 +238,10 @@ public Sensor addLatencySensor(final String scopeName,
         public void recordLatency(final Sensor sensor, final long startNs, final long endNs) {
 
         }
+
+        @Override
+        public Map<MetricName, ? extends Metric> metrics() {
+            return null;
```
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/StreamThreadStateStoreProviderTest.java:244]: As above. Maybe we need to have a `NoOpStreamsMetrics` class that can be used for tests?

File: streams/src/test/java/org/apache/kafka/test/MockProcessorContext.java [hunk 3]
```diff
@@ -141,6 +143,10 @@ public Sensor addLatencySensor(String scopeName, String entityName, String opera
             @Override
             public void recordLatency(Sensor sensor, long startNs, long endNs) {
             }
+            @Override
+            public Map<MetricName, ? extends Metric> metrics() {
+                return null;
```
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/test/MockProcessorContext.java:148]: And again - i'd say it is definitely worth having what i mentioned above

**Review by dguy — CHANGES_REQUESTED** [review]
Left some comments about having a `NoOpStreamMetrics` class that is used for testing. There is already four usages of it here, so i think it is worth adding.

**Comment by enothereska** [comment]
@dguy thanks I agree. I already have a MockStreamsMetrics in the other PR https://github.com/apache/kafka/pull/1446/. Is it ok if I use that PR to make the change above? Thanks.

**Comment by dguy** [comment]
@enothereska - of course.

**Comment by enothereska** [comment]
kafka.producer.ProducerTest > testAsyncSendCanCorrectlyFailWithTimeout FAILED 
Not PR related.

**Comment by guozhangwang** [comment]
I have a meta question for this PR and https://github.com/apache/kafka/pull/1446:

In #1446, we are adding three `sensor` APIs to the StreamsMetrics as well, and I'm wondering how that will be aligned with this PR that expose the inner general `Metrics` directly? I feel that we should choose to do either one of the two options:

1. Do not expose a general API in `StreamsMetrics` like the one exposed in #1446, and just expose the one in this PR to return the inner `? extends Metrics` to let users do whatever they want.
2. Expose the `sensor` API as proposed in #1446, which is an indirect way of adding any type of sensors (note that `StreamsMetrics` is accessible from `ProcessorContext`), with a bit of principles for specifying `scopeName`, `entityName`, and `operationName` etc, and how that will be translated to the logic of adding sensors to the inner metrics.

If we do both of these two set of APIs, it is quite confusing from a user's perspective.

**Comment by dguy** [comment]
@guozhangwang -  I see them as two separate concerns

1. I'd say exposing the Map of metrics collected by Streams is a good thing. It is consistent with the Consumer and Producer and also gives users a way of getting the metrics and doing stuff with them ( i have always found having to configure a metrics collector class as inconvenient ). So to me, this is more about - "hey give me access to the metrics that you have collected"

2. This seems more about letting users add their own metrics to `StreamMetrics`. I'm not 100% convinced in the value of doing this.  For the most part people will have using some other metrics libraries and collection mechanism. Generally, i think,  they will just want to get the metrics out of streams and plug them into their own mechanism.  So, i don't really think we should be exposing a Metrics API

**Comment by guozhangwang** [comment]
@dguy I was not reading the diff file carefully, and my intention was from the original description in the JIRA ticket itself:

```
If there were a "metrics()" method of StreamMetrics to give me the underlying Metrics object, I could register my own sensors/metrics as needed.
```

I think this PR's proposed APIs for exposing the read-only metrics map for users to lookup does makes sense and it is consistent with other clients. But I think users also have a need to register their own metrics by using the `StreamsMetrics` object. Maybe this second intention is already covered by #1446, which I am now sure if enforcing the naming on scope / entity / operation is a good idea or not, and we can discuss on that regard.

**Comment by guozhangwang** [comment]
@enothereska I think KAFKA-3537 was filed for exposing the metrics registry to allow users to register their own, while this PR is for exposing the metrics map in read-only manner as we did for other clients, i.e. it should be really targeting https://issues.apache.org/jira/browse/KAFKA-3701? If you agree, could we update the PR title then?

EDIT: thinking about this a bit more, I feel there are some further differences in KAFKA-3701 and KAFKA-3537: 

1. If we want to add the read-only metrics map as we did for other clients, for the purposes of health checking / building their own metrics mechanism based on this `getter` API, then it is better to be on the `KafkaStreams` class to return the `Map<MetricName, ? extends Metric>`. (KAFKA-3701).

2. If we want to allow users to registry their own metrics by leveraging the library's metric collector, then it is better to be added on the `StreamsMetrics` class, which returns the `Metrics` object (KAFKA-3537).

**Comment by enothereska** [comment]
This will all be part of this PR moving forward: https://github.com/apache/kafka/pull/1446

**Pull request closed without merging**